### DISCUSSION TITLE
Proper Cleanup - Version Change from 0.27.0 to 0.28.0

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#222] [Q&A] [answered]
I was updating my library to the latest 0.29.0, and noticed something that is either a change in the ShapeSettings behavior of the CPP side, or a change in how we are supposed to Release/Cleanup,  to be able to 100% clean up shapes.

Running the 'proper cleanup' with 0.27.0 results in a 0 byte difference (and I was able to reproduce this in my physics plugin using Jolt as full)
Running the 'proper cleanup' code with 0.28 results in something like 16-24 bytes extra per shape, even going through the full `shapeResult.Clear()`. It does not appear to be the same as the old cache, since passing in the full 'Falling Shapes - Mesh Ground' does not add more than the 16-24 bytes.

I created a Fiddle that allows for quickly testing this out and confirming that with Zero shapes, the memory clears to 0-bytes, and commenting out to 100% exclude the different Shapes reduces it by a fixed amount each. 1x Tapered Capsule leaves the same residual memory amount as 10x TaperedCapsule.
https://jsfiddle.net/phoenixillusion/zj5xo6ev/2/

Below is a minimal residual scenario, where the EmptyShape and SphereShape both appear to leak 16 bytes
```javascript
<!DOCTYPE html>
<html lang="en">
  <body>
    <script type="module">
      //import initJolt from 'https://unpkg.com/jolt-physics@0.27.0/dist/jolt-physics.wasm-compat.js';
      import initJolt from 'https://unpkg.com/jolt-physics@0.28.0/dist/jolt-physics.wasm-compat.js';

      initJolt().then(function (Jolt) {
        // Record how much memory we have in the beginning
        const memoryFreeBefore = Jolt.JoltInterface.prototype.sGetFreeMemory();

        {
          const shapeSettings = new Jolt.SphereShapeSettings(0.5);
          const shapeResult = shapeSettings.Create();
          const shape = shapeResult.Get();
          shape.AddRef();
          Jolt.destroy(shapeSettings);
          shape.Release();
          shapeResult.Clear();
        }
        {
          const shapeSettings = new Jolt.EmptyShapeSettings();
          const shapeResult = shapeSettings.Create();
          const shape = shapeResult.Get();
          shape.AddRef();
          Jolt.destroy(shapeSettings);
          shape.Release();
          shapeResult.Clear();
        }
        
        // Record how much memory we have at the end
        document.write("Memory leaked: " + (memoryFreeBefore - Jolt.JoltInterface.prototype.sGetFreeMemory()) + " bytes");
      });
    </script>
  </body>
</html>
```

I've tried tracking down what may creating the allocations.
I can see that in the following calls, the last call to Clear() releases 40 bytes, but the shapeSettings.Create() allocates 56 bytes, leaving a 16 byte residual.
In the case of "EmptyShapeSettings", the memory is instead 80 byte allocate and 56 byte free, so a 24 residual.

```javascript
        const mem = () => Jolt.JoltInterface.prototype.sGetFreeMemory();
        const memoryFreeBefore = mem();
        let x = memoryFreeBefore;
        const m = () => { const z = mem(); const ret = x - z; x = z; return ret; };
        {
          const shapeSettings = new Jolt.SphereShapeSettings(0.5);console.log(m())
          const shapeResult = shapeSettings.Create();console.log(m())
          const shape = shapeResult.Get();console.log(m())
          Jolt.destroy(shapeSettings);console.log(m());
          shapeResult.Clear();console.log(m())
         }
```

**Comment by jrouwe** [comment]
Hello,

This is the result of [this change in emscripten](https://github.com/emscripten-core/emscripten/pull/22772) (which I made myself). The multithreaded version of JoltPhysics.js requires thread safe access to temporaries. Before the generated code for `SphereShapeSettings.Create` looked like this:

```C++
Shape::ShapeResult* EMSCRIPTEN_KEEPALIVE emscripten_bind_SphereShapeSettings_Create_0(SphereShapeSettings* self) {
  static Shape::ShapeResult temp;
  return (temp = self->Create(), &temp);
}
```

now it looks like:

```C++
Shape::ShapeResult* EMSCRIPTEN_KEEPALIVE emscripten_bind_SphereShapeSettings_Create_0(SphereShapeSettings* self) {
  static thread_local Shape::ShapeResult temp;
  return (temp = self->Create(), &temp);
}
```

Creating a `thread_local` involves allocating some memory which will not be freed until the application terminates. This is why you see a small leak (which is not actually a leak) per method that uses a `thread_local` (e.g. all the `XXXShapeSettings.Create` functions). 

You can verify this yourself by removing the `thread_local` keyword from `glue.cpp` by editing `CMakeLists.txt` and changing:

```
add_custom_command(
	OUTPUT glue.cpp glue.js jolt.idl
	BYPRODUCTS parser.out WebIDLGrammar.pkl
	COMMAND cat ${JOLT_IDL_FILE} > jolt.idl
	COMMAND ${PYTHON} ${WEBIDL_BINDER_SCRIPT} jolt.idl glue
	DEPENDS ${JOLT_IDL_FILE}
	COMMENT "Generating JoltPhysics.js bindings"
	VERBATIM)
```

to

```
add_custom_command(
	OUTPUT glue.cpp glue.js jolt.idl
	BYPRODUCTS parser.out WebIDLGrammar.pkl
	COMMAND cat ${JOLT_IDL_FILE} > jolt.idl
	COMMAND ${PYTHON} ${WEBIDL_BINDER_SCRIPT} jolt.idl glue
	COMMAND sed -i s/thread_local//g glue.cpp
	DEPENDS ${JOLT_IDL_FILE}
	COMMENT "Generating JoltPhysics.js bindings"
	VERBATIM)
```

For the single threaded version this should work equally well, for the multi threaded version this will create a race condition and will make the program misbehave/crash.

**Reply by PhoenixIllusion** [reply]
I'd tried to stick to using the word 'residual' since it was clearly a cache-like leftover, although the original Proper Cleanup code used the word leak, so I did use that once. Now that it is confirmed as thread_local static, 'cache' would probably be a better name. Would that `sed` CMake command have any problems if it was tied to the `DENABLE_MULTI_THREADING=ON` command? I saw this was originally your intent on the PR to emscripten, allowing usage of a webidl_binder.py parameter controlling this value, that was later removed.

I understand the purpose, and saw the existing 'thread_local' change was clean enough for merging into the main emscripten branch, and is on emscripten master right now.
If the we did maintain the code as-is, without patching out thread_local on single-thread builds, can you think of a good way to memory audit? Measuring "back to zero bytes" was easy.

I am trying to think if there is a memory-audit work-around beyond just maintaining a custom builds to strip it out for testing purposes, given the alternative is just assuming after library tear-down one should "just assume if there are a couple 1000 bytes leftover, after using sufficient usage of the thread_local features, that they should be there". Just assuming that can cover up stray vec3 or other small items that got allocated in a special or randomized case of my code and forgotten, and these stray 4 bytes would be missed in triggering usual leak detection vs "back to zero bytes". 

I know another alternative is to make a 'priming' step of my library where I try to pre-fill all these thread_local slots, and use that as my new "zero bytes", but that requires a I can make a pristinely coded routine in my tests that can allocate every possible thread_local without leaking anything else, which in itself has risks of poor code.

Technically I could loop all test scenarios, grab the residual after cleaning the final scenario, and then re-run all memory audits tests a second time and use that final static-residual as my new 'zero' in the second loop of tests.

**Reply by PhoenixIllusion** [reply]
I've currently tried this with:
```
if (ENABLE_MULTI_THREADING)
    ...
    set(REMOVE_THREAD_LOCAL "")
else()
    ...
    set(REMOVE_THREAD_LOCAL sed -i.bak -e "s:thread_local::g" glue.cpp)
    # tried to keep this multi-platform since windows/mac/linux all treat -i different,
    # though creates extra backup file in a non-tracked folder
endif()

...

add_custom_command(
	OUTPUT glue.o
	COMMAND ${REMOVE_THREAD_LOCAL}
	COMMAND emcc glue.cpp ${EMCC_GLUE_ARGS} -o glue.o
```

**Comment by jrouwe** [comment]
I've integrated your change. Thanks!

**Reply by jrouwe** [reply, marked as answer]
I released the change as part of 0.30.0

**Reply by PhoenixIllusion** [reply]
Thank you. I was still trying to figure out any tricks to avoid pre-build file modification, since I always prefer working with the cleanest and unmodified source I can.

**Reply by jrouwe** [reply]
I did make a change compared to your version, it is applied after the webidl script runs, so technically it's not a pre-build file modification anymore.
